crates/vault-registry/src/lib.rs

```diff
@@ -308,38 +326,132 @@ impl<T: Config> Module<T> {
+        Ok(())
+    }
+
+    fn force_withdraw_collateral(vault_id: &T::AccountId, amount: DOT<T>) -> DispatchResult {
+        let mut vault = Self::get_rich_vault_from_id(vault_id)?;
+        vault.force_withdraw_collateral(amount)?;
+        Ok(())
+    }
+
+    pub fn slash_collateral_saturated(
+        from: CurrencyType<T>,
+        to: CurrencyType<T>,
+        amount: DOT<T>,
+    ) -> Result<DOT<T>, DispatchError> {
+        let available_amount = from.current_balance()?;
+        let amount = if available_amount < amount {
+            available_amount
+        } else {
+            amount
+        };
+        Self::slash_collateral(from, to, amount)?;
+        Ok(amount)
+    }
+
+    pub fn slash_collateral(
+        from: CurrencyType<T>,
+        to: CurrencyType<T>,
+        amount: DOT<T>,
+    ) -> DispatchResult {
+        // move funds to free balance of the source
+        match from {
+            CurrencyType::Backing(ref account) => {
+                Self::force_withdraw_collateral(account, amount)?;
+            }
+            CurrencyType::Griefing(_) | CurrencyType::LiquidationVault => {
+                ext::collateral::release_collateral::<T>(&from.account_id(), amount)?;
+            }
+            CurrencyType::FreeBalance(_) => {
+                // do nothing
+            }
+        };
+
+        // move from sender's free balance to receiver's free balance
+        ext::collateral::transfer::<T>(&from.account_id(), &to.account_id(), amount)?;
+
+        // move funds to free balance of the source
+        match to {
+            CurrencyType::Backing(ref account) => {
+                Self::_lock_additional_collateral(account, amount)?;
+            }
+            CurrencyType::Griefing(_) | CurrencyType::LiquidationVault => {
+                ext::collateral::lock::<T>(&to.account_id(), amount)?;
+            }
+            CurrencyType::FreeBalance(_) => {
+                // do nothing
+            }
+        };
+
+        Ok(())
+    }
+    /// Increases the amount of tokens to be issued (in replace or issue)
+    ///
+    /// # Arguments
+    /// * `vault_id` - the id of the vault from which to increase to-be-issued tokens
+    /// * `tokens` - the amount of tokens to be reserved
     pub fn increase_to_be_issued_tokens(
         vault_id: &T::AccountId,
-        secure_id: H256,
         tokens: PolkaBTC<T>,
-    ) -> Result<BtcAddress, DispatchError> {
+    ) -> Result<(), DispatchError> {
         ext::security::ensure_parachain_status_running::<T>()?;
         let mut vault = Self::get_active_rich_vault_from_id(&vault_id)?;
         vault.increase_to_be_issued(tokens)?;
         Self::deposit_event(Event::<T>::IncreaseToBeIssuedTokens(vault.id(), tokens));
-        let btc_address = vault.new_deposit_address(secure_id)?;
+        Ok(())
+    }
+
+    /// Registers a btc address
+    ///
+    /// # Arguments
+    /// * `issue_id` - secure id for generating deposit address
+    pub fn _register_address(
```

Comment:
Naming is confusing because the similarly named dispatchable function above, maybe we could rename to `register_deposit_address` unless you have another idea?